### PR TITLE
Update Drivetrain

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package frc.robot.Commands;
+
+import com.spikes2212.command.drivetrains.commands.DriveArcadeWithPID;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.Subsystems.Dispenser;
+import frc.robot.Subsystems.Drivetrain;
+
+
+public class AimToHub extends SequentialCommandGroup {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:10]: add documentations

File: src/main/java/frc/robot/Subsystems/Dispenser.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package frc.robot.Subsystems;
+
+
+import com.spikes2212.util.Limelight;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Dispenser extends SubsystemBase {
+    public Limelight limelight;
+
+    // With eager singleton initialization, any static variables/fields used in the
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Dispenser.java:10]: what?

File: src/main/java/frc/robot/Subsystems/Dispenser.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    // before the constructor is called when the "INSTANCE" variable initializes.
+
+    private final static Dispenser INSTANCE = new Dispenser();
+
+    public static Dispenser getInstance() {
+        return INSTANCE;
+    }
+
+    private Dispenser() {
+        // TODO: Set the default command, if any, for this subsystem by calling setDefaultCommand(command)
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Dispenser.java:21]: what?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.spikes2212.command.drivetrains.commands.DriveArcadeWithPID;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.Subsystems.Dispenser;
+import frc.robot.Subsystems.Drivetrain;
+
+/**
+ * <p>
+ * A command which causes the robot to rotate in place until it centers on one of the
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:11]: causes?

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * reflective sprites that are on the hub.
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:12]: that's not the right description

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * </p>
+ * @author Yoel Perman Brilliant
+ */
+public class AimToHub extends SequentialCommandGroup {
+    private static final int LIMELIGHT_PIPE = 1;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:17]: pipe?

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    public AimToHub() {
+        Dispenser dispenser = Dispenser.getInstance();
+        Drivetrain drivetrain = Drivetrain.getInstance();
+        addCommands(
+                new InstantCommand(() -> dispenser.limelight.setPipeline(LIMELIGHT_PIPE)),
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:23]: why command?

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                new DriveArcadeWithPID(drivetrain, dispenser.limelight::getHorizontalOffsetFromTarget,
+                        () -> 0.0, () -> 0.0, drivetrain.aimToHubPID, drivetrain.aimToHubFeedForward)
+        );
+    }
+
+
+}
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:30]: add blank line at the end of the file

File: src/main/java/frc/robot/Subsystems/Dispenser.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package frc.robot.Subsystems;
+
+
+import com.spikes2212.util.Limelight;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Dispenser extends SubsystemBase {
+    public Limelight limelight;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Dispenser.java:8]: private

File: src/main/java/frc/robot/Subsystems/Drivetrain.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package frc.robot.Subsystems;
+
+
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Drivetrain extends TankDrivetrain {
+    public PIDSettings aimToHubPID;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Drivetrain.java:10]: private + getter

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 11]
```diff
@@ -75,7 +75,7 @@ dependencies {
     simulationRelease wpi.sim.enableRelease()
 
     testImplementation 'junit:junit:4.13.2'
-    implementation 'com.github.Spikes-2212-Programming-Guild:SpikesLib2:master-SNAPSHOT'
+    implementation 'com.github.Spikes-2212-Programming-Guild:SpikesLib2:aaaaaaaah-SNAPSHOT'
```
Comment (tal-sitton) [inline build.gradle:78]: ?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 12]
```diff
@@ -1,5 +1,3 @@
-import edu.wpi.first.gradlerio.deploy.roborio.RoboRIO
```
Comment (tal-sitton) [inline build.gradle:1]: Why did you remove this?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import frc.robot.Subsystems.Transfer;
+
+/**
+ * <p>
+ * A command which centers the robot on one of the reflective sprites that are on the hub.
+ * </p>
+ *
+ * @author Yoel Perman Brilliant
+ */
+public class AimToHub extends SequentialCommandGroup {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:15]: change to extends DriveArcadeWithPID

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/ReleaseCargo.java [hunk 14]
```diff
@@ -1,4 +1,5 @@
 package frc.robot.Commands;
 
 public class ReleaseCargo {
+
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/ReleaseCargo.java:4]: why?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * <p>
+ * A command which centers the robot on one of the reflective sprites that are on the hub.
+ * </p>
+ *
+ * @author Yoel Perman Brilliant
+ */
+public class AimToHub extends DriveArcadeWithPID {
+    private static final int LIMELIGHT_PIPELINE = 1;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:15]: blank line at the start of the class

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    private final Transfer transfer;
+    private int initialPipeline;
+
+
+    public AimToHub(Drivetrain drivetrain, Transfer transfer) {
+       super(drivetrain, transfer.getLimelight()::getHorizontalOffsetFromTarget, 0, 0,
+               drivetrain.getPIDSettings(), drivetrain.getFFSettings());
+       this.transfer = transfer;
+    }
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:26]: remove blank line

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @author Yoel Perman Brilliant
+ */
+public class AimToHub extends DriveArcadeWithPID {
+    private static final int LIMELIGHT_PIPELINE = 1;
+
+    private final Transfer transfer;
+    private int initialPipeline;
+
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:20]: remove blank line

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
+package frc.robot.Commands;
+
+import com.spikes2212.command.drivetrains.commands.DriveArcadeWithPID;
+import frc.robot.Subsystems.Drivetrain;
+import frc.robot.Subsystems.Transfer;
+
+/**
+ * <p>
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:8]: remove paragraph tag

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+ * A command which centers the robot on one of the reflective sprites that are on the hub.
+ * </p>
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:10]: remove paragraph tag

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * @author Yoel Perman Brilliant
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:12]: remove author

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public class AimToHub extends DriveArcadeWithPID {
+    private static final int LIMELIGHT_PIPELINE = 1;
+
+    private final Transfer transfer;
+    private int initialPipeline;
+
+
+    public AimToHub(Drivetrain drivetrain, Transfer transfer) {
+       super(drivetrain, transfer.getLimelight()::getHorizontalOffsetFromTarget, 0, 0,
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:22]: you are using the getLimelight method a lot of times. Save it in a variable and use it instead of calling the getter (fix it on all of the getter calls in this class)

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/ReleaseCargo.java (hunk 14 above)
Comment (Yoel-Perman-Brilliant) [reply src/main/java/frc/robot/Commands/ReleaseCargo.java:4]: Idk must be a bug

File: src/main/java/frc/robot/Commands/AimToHub.java (hunk 21 above)
Comment (Yoel-Perman-Brilliant) [reply src/main/java/frc/robot/Commands/AimToHub.java:22]: Not possible at the getter in the super because you can't initialize a field before the super call, but did it for the other getters

File: src/main/java/frc/robot/Subsystems/Drivetrain.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+import com.spikes2212.command.drivetrains.TankDrivetrain;
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.util.PigeonWrapper;
+
+public class Drivetrain extends TankDrivetrain {
+    private PIDSettings pidSettings;
+    private FeedForwardSettings ffSettings;
+
+    private PigeonWrapper gyro;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/Drivetrain.java:13]: if you don't use it, delete it

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]
add gradle.yml

File: github/workflows/gradle.yml [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
+        uses: actions/setup-java@v2
+        with:
+          java-version: '11'
+          distribution: 'adopt'
+      - name: Grant execute permission for gradlew
+        run: chmod +x gradlew
+      - name: Build with Gradle
+        run: ./gradlew build --scan
```
Comment (OfriRosenbaum) [inline github/workflows/gradle.yml:20]: add a blank line

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+               drivetrain.getPIDSettings(), drivetrain.getFFSettings());
+       this.transfer = transfer;
+       this.limelight = transfer.limelight;
+    }
+
+    @Override
+    public void initialize() {
+        initialPipeline = limelight.getPipeline();
+        limelight.setPipeline(LIMELIGHT_PIPELINE);
+        super.initialize();
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/AimToHub.java:30]: why would you call this super method?

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    private static final int LIMELIGHT_PIPELINE = 1;
+    private static final RootNamespace aimToHubNS = new RootNamespace("aim to hub");
+    private static final Supplier<Double> MOVE_VALUE = aimToHubNS.addConstantDouble("move value", 0);
+
+    private final Transfer transfer;
+    private int initialPipeline;
+    private final Limelight limelight;
+
+    public AimToHub(Drivetrain drivetrain, Transfer transfer) {
+       super(drivetrain, transfer.getLimelight()::getHorizontalOffsetFromTarget, 0, MOVE_VALUE.get(),
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:25]: why `.get()`

File: src/main/java/frc/robot/Commands/AimToHub.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * A command which centers the robot on one of the reflective sprites that are on the hub.
+ */
+public class AimToHub extends DriveArcadeWithPID {
+
+    private static final int LIMELIGHT_PIPELINE = 1;
+    private static final RootNamespace aimToHubNS = new RootNamespace("aim to hub");
+    private static final Supplier<Double> MOVE_VALUE = aimToHubNS.addConstantDouble("move value", 0);
+
+    private final Transfer transfer;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/AimToHub.java:20]: why?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotMap.java [hunk 27]
```diff
@@ -1,4 +1,37 @@
...
+        int DRIVETRAIN_RIGHT_ENCODER_NEG = 7;
+
+        int INTAKE_PLACER_UPPER_LIMIT = 0;
+        int INTAKE_PLACER_LOWER_LIMIT = 1;
+
+        int TRANSFER_ENTRANCE_LIGHT_SENSOR = 2;
+
+        int INTAKE_TO_TRANSFER_LIMIT = 3;
+    }
+}
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java:37]: add a blank line

File: src/main/java/frc/robot/Subsystems/Drivetrain.java [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+        });
+        gyroNamespace.putNumber("yaw", this::getYaw);
+        gyroNamespace.putData(" pigeon", new InstantCommand(pigeon::reset) {
+            @Override
+            public boolean runsWhenDisabled() {
+                return true;
+            }
+        });
+    }
+}
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Subsystems/Drivetrain.java:198]: add a blank line

File: vendordeps/SpikesLib2.json [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+    "cppDependencies": [
+    ],
+    "javaDependencies": [
+        {
+            "groupId": "com.github.Spikes-2212-Programming-Guild",
+    	    "artifactId": "SpikesLib2",
+    	    "version": "v2.0.3"
+        }
+    ]
+}
```
Comment (OfriRosenbaum) [inline vendordeps/SpikesLib2.json:20]: add a blank line

File: src/main/java/frc/robot/commands/Transit.java [hunk 30]
```diff
@@ -0,0 +1,4 @@
+package frc.robot.commands;
+
+public class Transit {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Transit.java:3]: delete this class

File: src/main/java/frc/robot/commands/Climb.java [hunk 31]
```diff
@@ -0,0 +1,4 @@
+package frc.robot.commands;
+
+public class Climb {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/Climb.java:3]: delete this class

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 32]
```diff
@@ -0,0 +1,4 @@
+package frc.robot.subsystems;
+
+public class Climber {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Climber.java:3]: delete this class

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 33]
```diff
@@ -0,0 +1,4 @@
+package frc.robot.subsystems;
+
+public class Intake {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/Intake.java:3]: delete this class

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 34]
```diff
@@ -19,11 +20,20 @@
 
 public class Drivetrain extends TankDrivetrain {
 
+    /**
+     * The wheel moves 15.24 * PI (it's perimeter) each 360 ticks (in meters).
+     */
+    private static final double DISTANCE_PER_PULSE = 15.24 * Math.PI / 360.0 / 100;
+    public static final int LIMELIGHT_PIPELINE = 1;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:27]: remvove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 35]
```diff
@@ -43,13 +55,27 @@
...
+    private final PIDSettings pidSettingsEncoders;
+
     private final Supplier<Double> kPCamera = cameraPIDNamespace.addConstantDouble("kP", 0.0065);
     private final Supplier<Double> kICamera = cameraPIDNamespace.addConstantDouble("kI", 0);
     private final Supplier<Double> kDCamera = cameraPIDNamespace.addConstantDouble("kD", 0);
     private final Supplier<Double> toleranceCamera = cameraPIDNamespace.addConstantDouble("tolerance", 2);
     private final Supplier<Double> waitTimeCamera = cameraPIDNamespace.addConstantDouble("wait time", 999);
     private final PIDSettings pidSettingsCamera;
 
+    private final Supplier<Double> kPLimelight = limelightPIDNamespace.addConstantDouble("kP", 0.1);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:72]: 0.1 is too much

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 36]
```diff
@@ -43,13 +55,27 @@
...
+    private final Supplier<Double> kILimelight = limelightPIDNamespace.addConstantDouble("kI", 0);
+    private final Supplier<Double> kDLimelight = limelightPIDNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> toleranceLimelight = limelightPIDNamespace.addConstantDouble("tolerance", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:75]: add some tolerance

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 37]
```diff
@@ -43,13 +59,27 @@
...
+    private final PIDSettings pidSettingsEncoders;
+
     private final Supplier<Double> kPCamera = cameraPIDNamespace.addConstantDouble("kP", 0.0065);
     private final Supplier<Double> kICamera = cameraPIDNamespace.addConstantDouble("kI", 0);
     private final Supplier<Double> kDCamera = cameraPIDNamespace.addConstantDouble("kD", 0);
     private final Supplier<Double> toleranceCamera = cameraPIDNamespace.addConstantDouble("tolerance", 2);
     private final Supplier<Double> waitTimeCamera = cameraPIDNamespace.addConstantDouble("wait time", 999);
     private final PIDSettings pidSettingsCamera;
 
+    private final Supplier<Double> kPLimelight = limelightPIDNamespace.addConstantDouble("kP", 0.0133);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:76]: remove all this, the PID for the limelight is the same as the cargoCam

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — CHANGES_REQUESTED** [review]
